### PR TITLE
Dedup UnboundedChannel and UnboundedPriorityChannel

### PR DESCRIPTION
**Pull request by stephentoub**
We can use generic specialization to avoid duplicating all the code for the different queue types. This should also make it much simpler to add other queue types in the future.

### CODE REVIEW

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs [hunk 1]
```diff
@@ -112,24 +114,60 @@ public override ValueTask<T> ReadAsync(CancellationToken cancellationToken)
...
             }
 
-            public override bool TryPeek([MaybeNullWhen(false)] out T item) =>
-                _parent._items.TryPeek(out item);
+            public override bool TryPeek([MaybeNullWhen(false)] out T item)
+            {
+                UnboundedChannel<T, TQueue> parent = _parent;
+                return parent._items.IsThreadSafe ?
+                    parent._items.TryPeek(out item) :
+                    NotIsThreadSafe(parent, out item);
```
Comment (stephentoub) [inline src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs:155]: This also fixes a bug in my previous PR that added the prioritized channel: TryPeek wasn't appropriately locking.

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/Channel.netcoreapp.cs [hunk 2]
```diff
@@ -2,20 +2,22 @@
...
 using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
 
 namespace System.Threading.Channels
 {
     /// <summary>Provides static methods for creating channels.</summary>
     public static partial class Channel
     {
         /// <summary>Creates an unbounded prioritized channel usable by any number of readers and writers concurrently.</summary>
+        /// <typeparam name="T">Specifies the type of data in the channel.</typeparam>
```
Comment (adamsitnik) [inline src/libraries/System.Threading.Channels/src/System/Threading/Channels/Channel.netcoreapp.cs:13]: thanks! 👍

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/IDebugEnumerator.cs [hunk 3]
```diff
@@ -26,4 +26,6 @@ public DebugEnumeratorDebugView(IDebugEnumerable<T> enumerable)
         [DebuggerBrowsable(DebuggerBrowsableState.RootHidden)]
         public T[] Items { get; }
     }
+
+    internal sealed class DebugEnumeratorDebugView<T, TOther>(IDebugEnumerable<T> enumerable) : DebugEnumeratorDebugView<T>(enumerable);
```
Comment (adamsitnik) [inline src/libraries/System.Threading.Channels/src/System/Threading/Channels/IDebugEnumerator.cs:30]: just for my education: why do we need `TOther` if it's not used by the type?

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/Channel.netcoreapp.cs [hunk 4]
```diff
@@ -30,7 +32,45 @@ public static Channel<T> CreateUnboundedPrioritized<T>(UnboundedPrioritizedChann
         {
             ArgumentNullException.ThrowIfNull(options);
 
-            return new UnboundedPrioritizedChannel<T>(!options.AllowSynchronousContinuations, options.Comparer);
+            return new UnboundedChannel<T, PriorityQueueUnboundedChannelQueue<T>>(new(new(options.Comparer)), !options.AllowSynchronousContinuations);
+        }
+
+        /// <summary>Provides an <see cref="IUnboundedChannelQueue{T}"/> for a <see cref="PriorityQueue{TElement, TPriority}"/>.</summary>
+        private readonly struct PriorityQueueUnboundedChannelQueue<T>(PriorityQueue<bool, T> queue) : IUnboundedChannelQueue<T>
```
Comment (adamsitnik) [inline src/libraries/System.Threading.Channels/src/System/Threading/Channels/Channel.netcoreapp.cs:39]: How about renaming these types to:

```diff
- ConcurrentQueueUnboundedChannelQueue
+ UnboundedChannelConcurrentQueue
- PriorityQueueUnboundedChannelQueue
+ UnboundedChannelPriorityQueue
```

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/Channel.netcoreapp.cs [hunk 5]
```diff
@@ -30,7 +32,45 @@ public static Channel<T> CreateUnboundedPrioritized<T>(UnboundedPrioritizedChann
...
+            /// <inheritdoc/>
+            public int Count => _queue.Count;
+
+            /// <inheritdoc/>
+            public bool IsEmpty => _queue.Count == 0;
+
+            /// <inheritdoc/>
+            public IEnumerator<T> GetEnumerator()
+            {
+                List<T> list = [];
```
Comment (adamsitnik) [inline src/libraries/System.Threading.Channels/src/System/Threading/Channels/Channel.netcoreapp.cs:64]: Have you considered special casing empty queue here?

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs [hunk 6]
```diff
@@ -112,24 +114,60 @@ public override ValueTask<T> ReadAsync(CancellationToken cancellationToken)
...
+                        return true;
+                    }
+
+                    item = default;
+                    return false;
                 }
 
-                item = default;
-                return false;
+                static bool NotIsThreadSafe(UnboundedChannel<T, TQueue> parent, [MaybeNullWhen(false)] out T item)
```
Comment (adamsitnik) [inline src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs:134]: again, another subjective rename proposal

```suggestion
                static bool NotThreadSafe(UnboundedChannel<T, TQueue> parent, [MaybeNullWhen(false)] out T item)
```

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs [hunk 7]
```diff
@@ -112,24 +114,60 @@ public override ValueTask<T> ReadAsync(CancellationToken cancellationToken)
 
             public override bool TryRead([MaybeNullWhen(false)] out T item)
             {
-                UnboundedChannel<T> parent = _parent;
+                UnboundedChannel<T, TQueue> parent = _parent;
+                return parent._items.IsThreadSafe ?
+                    IsThreadSafe(parent, out item) :
+                    NotIsThreadSafe(parent, out item);
```
Comment (adamsitnik) [inline src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs:120]: again, another subjective rename proposal

```suggestion
                    ThreadSafe(parent, out item) :
                    NotThreadSafe(parent, out item);
```

some other ideas: `Locked` and `LockFree`

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs (hunk 1 above)
Comment (adamsitnik) [reply src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs:155]: Thanks for adding the comment! FWIW you have marked it as resolved and I almost missed it:

image

**Review by adamsitnik — APPROVED** [review]
LGTM!

I've left some comments but they were all subjective rename proposals, so feel free to ignore them or address after merging this PR before Preview 4 snap (solving the bug is more important ;) ).

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/IDebugEnumerator.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Threading.Channels/src/System/Threading/Channels/IDebugEnumerator.cs:30]: This type is used as a DebuggerTypeProxy. The number of generic type parameters for the proxy needs to match the number for the type it's proxying.

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs (hunk 7 above)
Comment (stephentoub) [reply src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs:120]: Renamed. Thanks.

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/Channel.netcoreapp.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/System.Threading.Channels/src/System/Threading/Channels/Channel.netcoreapp.cs:39]: Renamed. Thanks.

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Threading.Channels/src/System/Threading/Channels/UnboundedChannel.cs:155]: Sorry :)

File: src/libraries/System.Threading.Channels/src/System/Threading/Channels/IDebugEnumerator.cs (hunk 3 above)
Comment (adamsitnik) [reply src/libraries/System.Threading.Channels/src/System/Threading/Channels/IDebugEnumerator.cs:30]: > The number of generic type parameters for the proxy needs to match the number for the type it's proxying.

Thanks!

**Pull request merged**